<a href="https://colab.research.google.com/github/IsabellaHerrerae/Anal-tica-/blob/main/1_NaiveBayes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Sección nueva

**Modelo de preaprobación - Naive Bayes**
0. Se procede con ña carga de las librerias de trabajo


In [ ]:
import numpy as np #Funciones Matemáticas
import pandas as pd #Comunicación con Excel
from sklearn.naive_bayes import GaussianNB #Función Gaussiana
from sklearn.metrics import confusion_matrix #Métricas de medida




1. Se cargan los datos


In [ ]:
nxl='/content/0. SolicitantesCrédito(USD).xlsx'
XDB=pd.read_excel(nxl,sheet_name=2) #Traigo los datos
XDB.dropna() #eliminat datos faltantes, celdas vacias

#Seleccionar las variables de interés
XDB=XDB[['Edad','Ingresos','Egresos','Cuota (COP)','PreApr']]


#Datos de entrada - np.array() - convertir a numeros
XD=np.array(XDB[['Edad','Ingresos','Egresos','Cuota (COP)']])
#Datos de salida -valor deseado - que quiero que aprenda el modelo
yd=np.array(XDB['PreApr'])

print(XDB) #Mostart la base de datos



      Edad    Ingresos      Egresos  Cuota (COP)  PreApr
0       38  1356.14400  1685.622667    18.375890       1
1       51   286.01600   364.462000     5.062492       1
2       37   517.46325   629.208889    10.287688       0
3       29   473.27000   551.616889    15.458878       1
4       42   750.09175   806.715778    24.995180       1
...    ...         ...          ...          ...     ...
5837    48  1207.84800   753.801111    27.043503       1
5838    31  1472.77200   953.812889    25.579568       1
5839    38   773.01975   672.910667    21.508792       1
5840    43   635.50175   780.691556    11.047468       1
5841    37  1419.26675  1169.838444    36.356182       1

[5842 rows x 5 columns]


2. Implementos Modelo

In [ ]:
mnb=GaussianNB() #Llamamos el modelo
mnb.fit(XD,yd) #Entrenamos el modelo
ydp=mnb.predict(XD) #Muestra los datos de salida - pero que el obtuvo

#Digame las medias de las variables para cada categoria - PreApr/PreNeg
u=mnb.theta_ #Se estiman las medidas
sigma=np.sqrt(mnb.var_) #Se estiman las desviaciones

#Las variaciones del las variables
LI=u-sigma
LS=u+sigma

print('La media de las variables por categoría:')
print(u)
print('El limite inferior de la variación de las variables por categoria es:')
print(LI)
print('El limite superior de la variación de las variables por categoria es:')
print(LS)


#Métrica de evaluación - cómo se comportó el modelo identificando la PreApr-PreNeg
cm=confusion_matrix(yd,ydp)
print('La matriz de confusión es:')
print(cm)

La media de las variables por categoría:
[[  40.98411626  431.42680644  401.54995644   16.88529983]
 [  41.10752688 1029.98947251  900.19312144   43.3269839 ]]
El limite inferior de la variación de las variables por categoria es:
[[ 28.77108358 105.62889147  58.77709557 -12.88412769]
 [ 28.85646196 557.55604444 374.75974307 -21.65436704]]
El limite superior de la variación de las variables por categoria es:
[[  53.19714893  757.22472141  744.32281731   46.65472735]
 [  53.35859181 1502.42290058 1425.62649981  108.30833484]]
La matriz de confusión es:
[[2563  396]
 [1192 1691]]


3. Se procede con la evaluación del modelo

In [ ]:
VN=cm[0,0] #Verdaderos negativos - Que tanto acierto tuvo el modelo frente a los que se les negó el crédito
FP=cm[0,1] #Falsos positivos
FN=cm[1,0] #Falso negativo -
VP=cm[1,1] #Verdadero positivo

Exactitud=(VN+VP)/len(XD) #len(XD) es el total de los datos
print(' La Exactitud es:', Exactitud) #comportamiento general


TE=(FN+FP)/len(XD)
print('El Tasa de Error es:',TE)

Sensibilidad=VP/(VP+FP)
print('La Sensibilidad es:',Sensibilidad) #Como se comporto frente a los positivos

Especificidad=VN/(VN+FN)
print('La Especificidad es:',Especificidad) #Como se comportaron frente a los negativos

Precision=VP/(VP+FN)
print('La Precisión es:',Precision) #cómo le pego a clasificados positivos

PredNeg=VN/(VN+FP)
print('La probabilidad de negativo es:',PredNeg) #como le pegó a los negativos

 La Exactitud es: 0.7281752824375214
El Tasa de Error es: 0.2718247175624786
La Sensibilidad es: 0.8102539530426449
La Especificidad es: 0.6825565912117177
La Precisión es: 0.5865417967395075
La probabilidad de negativo es: 0.8661710037174721


4. Qué pasa con una sola persona

In [ ]:
Xun=np.array([[38,1356,1686,18]]) #Esta es la persona del Excel
ydun=mnb.predict(Xun) #Muestra los datos de salida - pero que el obtuvo
print(ydun)

[1]


Análisis de resultados:

de los resultados arrojados por el modelo, se puede observar que los preaprobados en general presentan edades (), ingresos() , egresos() y cuotas mayores a los aprobados. Con respecto a las métricas, se puede observar que el modelo funcionó mejor clasificando los positivos, como dice el indicador de sensibilidad el cual alcanzó el 81% de certeza. De manera general se puede ibsevar que el modelo logró una exactitud que se ubicó por encima del 70% en promedio, y una tasa de error al 30% en promedio. De manera general, esta exactitud lo convierte en un buen modelo. Con respecto al solicitante de crédito tal y tal este pertenece a la categoria de aprobado.